main/src/ca/uwaterloo/flix/language/phase/Inliner.scala

```diff
@@ -34,20 +40,350 @@ import ca.uwaterloo.flix.util.Validation._
+      }
+    }
 
-    root.toSuccess
+    exp match {
+      /* Inline application */
+      case ApplyRef(sym, args, _, _) =>
+        definitions(sym) match {
+          case Constant(_, _, formals, exp1, _, _, _) =>
+            if (scores(sym) <= Score.MaxScore){
+              val freshExp1 = FreshenExpr.freshenExpr(exp1)
+              val (revArgs, argSymbols) = argsToSym(args)
+              val formalMap = formals.map(f => f.sym).zip(argSymbols).toMap
+              val substituted = substituteWith(formalMap,freshExp1)
+              val letBoundExpr = letBindAndInlineArgs(revArgs, substituted)
+              Some(letBoundExpr)
+            }
+            else {
+              None
+            }
+        }
+      /* Inline inside expression */
+      case MkClosureRef(_, _, _, _) => None
+      case _: LoadExpression => None
+      case _: StoreExpression => None
+      case Unit => None
+      case True => None
+      case False => None
+      case Char(_) => None
+      case Float32(_) => None
+      case Float64(_) => None
+      case Int8(_) => None
+      case Int16(_) => None
+      case Int32(_) => None
+      case Int64(_) => None
+      case BigInt(_) => None
+      case Str(_) => None
+      case Var(_, _, _) => None
+      case Ref(_, _, _) => None
+      case Lambda(_, _, _, _) => None
+      case Hook(_, _, _) => None
+      case ApplyTail(_, _, _, _, _) => None
+      case ApplyHook(_, _, _, _) => None
+      case Apply(_, _, _, _) => None
+      case Unary(_, _, _, _) => None
+      case Binary(_, _, _, _, _) => None
+      case IfThenElse(_, _, _, _, _) => None
+      case Let(_, _, _, _, _) => None
+      case LetRec(_, _, _, _, _) => None
+      case Is(_, _, _, _) => None
+      case Tag(_, _, _, _, _) => None
+      case Untag(_, _, _, _, _) => None
+      case Index(_, _, _, _) => None
+      case Tuple(_, _, _) => None
+      case Existential(_, _, _) => None
+      case Universal(_, _, _) => None
+      case NativeConstructor(_, _, _, _) => None
+      case NativeField(_, _, _) => None
+      case NativeMethod(_, _, _, _) => None
+      case UserError(_, _) => None
+      case MatchError(_, _) => None
+      case SwitchError(_, _) => None
+      /* Error */
+      case MkClosure(_, _, _, _) =>
+        throw InternalCompilerException(s"Unexpected expression $exp after lambda lifting")
+    }
+  }
+
+  /**
+    * Traverses through `exp` and performs inlining using
+    * definitions and sizes from `scores`
+    */
+  def inlineExpr(definitions: Map[DefnSym, Constant], scores: Map[DefnSym, Int], exp: Expression)(implicit genSym: GenSym): Expression =
+  {
+    def f = inlineVisitor(definitions,scores,_: Expression)
+    AstVisitor.visitExpressionWith(f, exp)
   }
 
+}
+
+object FreshenExpr {
+
+  /**
+    * Freshen all let binds in an expression
+    */
+  def freshenExpr(exp: Expression)(implicit genSym: GenSym): Expression =
+  {
+    val (m,newLets) = freshSymbols(exp)
+    Inliner.substituteWith(m,newLets)
```

Comment:
This seems to freshen (alpha rename) variables in an expression using two traversals (and the visitors!).

A much simpler solution is a single straightforward recursion on the expression which carries a substitution map (initially empty).